src/afl-fuzz-queue.c

```diff
@@ -767,3 +768,89 @@ u32 calculate_score(afl_state_t *afl, struct queue_entry *q) {
 
 }
 
+/* Tell afl that this testcase may be evicted from the cache */
+inline void queue_testcase_release(afl_state_t *afl, struct queue_entry *q) {
+
+  (void)afl;
+  if (unlikely(q->testcase_refs == 0)) {
+
+    FATAL("Testcase refcount reduced past 0");
+
+  }
+
+  q->testcase_refs--;
+
+}
+
+/* Returns the testcase buf from the file behind this queue entry.
+  Increases the refcount. */
+u8 *queue_testcase_take(afl_state_t *afl, struct queue_entry *q) {
+
+  if (!q->testcase_buf) {
+
+    u32 tid = 0;
+    /* Buf not cached, let's do that now */
+
+    if (likely(afl->q_testcase_cache_count == TESTCASE_CACHE_SIZE)) {
```

Comment:
for current afl-fuzz this is a bad approach. better: as afl-fuzz selects queue entries from start to finish. the ones should be evicted that have just been completed. The removes the need for `do { rand } while;` and instead you go backwards from where you currently are (or switch to end of queue if begin of queue is reached) to evict one entry

Reply:
I don't really agree here: most of the time the cache will be used for splicing, and splicing takes a random entry each time, so random eviction is fine. Going backwards also needs additional logic (granted, it's likely as fast or faster than rand, though)

Reply:
The while will still be needed, as the one before may already be in use from fuzzing during the splicing

Reply:
yes the while is needed. and as I said elsewhere it is a small performance thing. statistically it will be better. but that effort should only be done after it shows on fuzzbench that it helps performance. it should. fuzzbench shows otherwise, but this is because of the trim issue (I guess and hope). and we can increase the cache it will also help.
the env var also is something that would be for after it has proven to be useful and not now for another benchmark